R/data.table.R

```diff
@@ -1620,8 +1620,13 @@ replace_dot_alias = function(e) {
         (jsub[[1L]] == "[" ||
            (jsub[[1L]] == "[[" && is.name(jsub[[2L]]) && eval(call('is.atomic', jsub[[2L]]), x, parent.frame()))) &&
         (is.numeric(jsub[[3L]]) || jsub[[3L]] == ".N")
-      headopt = jsub[[1L]] == "head" || jsub[[1L]] == "tail"
-      firstopt = jsub[[1L]] == "first" || jsub[[1L]] == "last" # fix for #2030
+      narm_arg = function(fun, jsub) {
+        cal = match.call(fun, jsub)
+        isTRUE(cal[["na.rm"]])
+      }
+      # only lapply optimize if first/last has na.rm=FALSE see also #5168
+      headopt =  jsub[[1L]] == "head"  || jsub[[1L]] == "tail"
+      firstopt = jsub[[1L]] == "first" || jsub[[1L]] == "last" && !narm_arg(first, jsub) ## fix for #2030
```

Comment:
iiuc, the `||` part here needs parens around it. and if so, there's a test missing because `first(, na.rm=TRUE)` is still being optimized but that's not intended. Will do ...